### PR TITLE
Remove global counters in tests

### PR DESCRIPTION
**Pull request by justinchuby**
Remove global counters in tests from reference_evaluator_backend_test and automatic_upgrade_test to allow parallel test execution.

Run tests in parallel in CI.

### CODE REVIEW

**Review by jcwchen — COMMENTED** [review]
I guess the original test stats will be covered by the new codecov: https://github.com/onnx/onnx/pull/5660 instead?

**Comment by justinchuby** [comment]
> I guess the original test stats will be covered by the new codecov: #5660 instead?

Yep!

**Comment by justinchuby** [comment]
Need to test the notebooks separately because they are not parallelizable.

**Comment by justinchuby** [comment]
OK we need --dist loadscope

File: .azure-pipelines/Linux-CI.yml [hunk 1]
```diff
@@ -81,7 +81,7 @@ jobs:
   - script: |
       source venv/bin/activate
 
-      pytest -sv --cov=onnx --cov-report=xml --cov-append --cov-branch --junit-xml pytest.xml
+      pytest -sv --cov=onnx --cov-report=xml --cov-append --cov-branch --junit-xml pytest.xml -n auto
```
Comment (justinchuby) [inline .azure-pipelines/Linux-CI.yml:84]:
```suggestion
      pytest -sv --cov=onnx --cov-report=xml --cov-append --cov-branch --junit-xml pytest.xml -n auto --dist loadscope
```

File: .azure-pipelines/MacOS-CI.yml [hunk 2]
```diff
@@ -73,7 +73,7 @@ jobs:
     displayName: 'Install dependencies and ONNX'
 
   - script: |
-      pytest
+      pytest -n auto
```
Comment (justinchuby) [inline .azure-pipelines/MacOS-CI.yml:76]:
```suggestion
      pytest -n auto --dist loadscope
```

File: .azure-pipelines/MacOS-CI.yml [hunk 3]
```diff
@@ -83,7 +83,7 @@ jobs:
       export ORT_MAX_IR_SUPPORTED_VERSION=8
       export ORT_MAX_ML_OPSET_SUPPORTED_VERSION=3
       export ORT_MAX_ONNX_OPSET_SUPPORTED_VERSION=19
-      pytest
+      pytest -n auto
```
Comment (justinchuby) [inline .azure-pipelines/MacOS-CI.yml:86]:
```suggestion
      pytest -n auto --dist loadscope
```

File: .azure-pipelines/Windows-CI.yml [hunk 4]
```diff
@@ -72,7 +72,7 @@ jobs:
       )
 
       pip install -e ".[reference]" -v
-      pytest
+      pytest -n auto
```
Comment (justinchuby) [inline .azure-pipelines/Windows-CI.yml:75]:
```suggestion
      pytest -n auto --dist loadscope
```

File: .azure-pipelines/Windows-CI.yml [hunk 5]
```diff
@@ -118,7 +118,7 @@ jobs:
         pip uninstall -y onnx
         pip install .
 
-        pytest
+        pytest -n auto
```
Comment (justinchuby) [inline .azure-pipelines/Windows-CI.yml:121]:
```suggestion
        pytest -n auto --dist loadscope
```

**Comment by justinchuby** [comment]
cc @xadupre

**Review by jcwchen — COMMENTED** [review]
Thanks for `pytest -n auto --dist loadscope`, but I am surprise that the CI running time seems similar.

**Comment by justinchuby** [comment]
> Thanks for `pytest -n auto --dist loadscope`, but I am surprise that the CI running time seems similar.

The Python tests don’t take much time so we may not see a big difference. Nevertheless enabling parallel testing allows us to catch undesirable order dependencies or global state dependencies in our tests

File: onnx/test/version_converter/automatic_upgrade_test.py [hunk 6]
```diff
@@ -1738,32 +1735,6 @@ def test_RegexFullMatch(self) -> None:
...
-            "Optional",
-            "OptionalGetElement",
-            "OptionalHasElement",
-            "StringSplit",
-        ]
-        all_op_names = [op for op in all_op_names if op not in excluded_ops]
-
-        untested_ops = set(all_op_names) - set(tested_ops)
-        print(untested_ops)
-        assert len(untested_ops) == 0
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnx/test/version_converter/automatic_upgrade_test.py:1765]: Good point - any suggestions?

File: onnx/test/version_converter/automatic_upgrade_test.py [hunk 7]
```diff
@@ -14,12 +14,14 @@
 # to the most recent version and runs checker and shape inference on the final upgraded model.
 ####################################################################################
 
-tested_ops = []
-
 
 class TestAutomaticUpgrade(automatic_conversion_test_base.TestAutomaticConversion):
+    @classmethod
+    def setUpClass(cls):
+        self.tested_ops = []
```
Comment (justinchuby) [inline onnx/test/version_converter/automatic_upgrade_test.py:21]:
```suggestion
        cls.tested_ops = []
```

File: onnx/test/version_converter/automatic_upgrade_test.py [hunk 8]
```diff
@@ -1757,12 +1760,11 @@ def test_ops_tested(self) -> None:
...
             "StringSplit",
-        ]
-        all_op_names = [op for op in all_op_names if op not in excluded_ops]
+        }
+        expected_tested_ops = all_op_names - excluded_ops
 
-        untested_ops = set(all_op_names) - set(tested_ops)
-        print(untested_ops)
-        assert len(untested_ops) == 0
+        untested_ops = expected_tested_ops - set(tested_ops)
```
Comment (justinchuby) [inline onnx/test/version_converter/automatic_upgrade_test.py:1766]:
```suggestion
        untested_ops = expected_tested_ops - set(self.tested_ops)
```

File: onnx/test/version_converter/automatic_upgrade_test.py [hunk 9]
```diff
@@ -1757,12 +1760,11 @@ def test_ops_tested(self) -> None:
...
-        ]
-        all_op_names = [op for op in all_op_names if op not in excluded_ops]
+        }
+        expected_tested_ops = all_op_names - excluded_ops
 
-        untested_ops = set(all_op_names) - set(tested_ops)
-        print(untested_ops)
-        assert len(untested_ops) == 0
+        untested_ops = expected_tested_ops - set(self.tested_ops)
+        self.assertEqual(untested_ops, {})
```
Comment (justinchuby) [inline onnx/test/version_converter/automatic_upgrade_test.py:1767]:
```suggestion
        self.assertEqual(untested_ops, set())
```

**Review by gramalingam — APPROVED** [review]

**Pull request merged**